pvlib/spectrum/mismatch.py

```diff
@@ -1100,3 +1101,95 @@ def qe_to_sr(qe, wavelength=None, normalize=False):
+    Notes
+    -----
+    The average photon energy (APE) is an index used to characterise the solar
+    spectrum. It has been used widely in the Physics literature since the
+    1900s, but its application for solar spectral irradiance characterisation
+    in the context of PV performance modelling was proposed in [1]_. The APE
+    is calculated based on the principle that a photon's wavelength is
+    inversely proportional to its energy:
+
+    .. math::
+
+        E_\gamma = \frac{hc}{\lambda},
+
+    where :math:`E_\gamma` is the energy of a photon with wavelength
+    :math:`\lambda`, :math:`h` is the Planck constant, and :math:`c` is the
+    speed of light. Therefore, the average energy of all photons within a
+    single spectral irradiance distribution provides an indication of the
+    general shape of the spectrum. A higher average photon energy
+    (shorter wavelength) indicates a blue-shifted spectrum, while a lower
+    average photon energy (longer wavelength) would indicate a red-shifted
+    spectrum. This value of the average photon energy can be calculated by
+    dividing the total number of photons in the spectrum by the total energy in
+    the spectrum as follows [1]_:
+
+    .. math::
+
+        \varphi = \frac{1}{q} \cdot \frac{\int_a^b \Phi_\lambda \, d\lambda}
+        {\int_a^b E_\lambda \, d\lambda}.
+
+    :math:`\Phi_\lambda` is the photon flux density as a function of
+    wavelength, :math:`q` is the elementary charge used here so that the
+    average photon energy, :math:`\varphi`, is expressed in electronvolts (eV).
+    The integration limits, :math:`a` and :math:`b`, define the wavelength
+    range within which the APE is calculated. By default, this function
+    calculates the value for APE based on full wavelength range of the
+    ``spectral_irr`` parameter.
+
+    References
+    ----------
+    .. [1] Jardine, C., et al., 2002, January. Influence of spectral effects on
+       the performance of multijunction amorphous silicon cells. In Proc.
+       Photovoltaic in Europe Conference (pp. 1756-1759).
+
+    """
+
+    si = spectral_irr
+
+    if not isinstance(si, (pd.Series, pd.DataFrame)):
+        raise TypeError('`spectral_irr` must be either a'
+                        ' pandas Series or DataFrame')
+    # check si type
+
+    if (si < 0).any().any():
+        raise ValueError('Spectral irradiance data must be positive')
+    # check if si contains any negative irradiance values
+
+    hclambda = pd.Series((constants.h*constants.c)/(si.T.index*1e-9))
+    hclambda.index = si.T.index  # set wavelength as the index
+    pfd = si.div(hclambda)  # calculate the photon flux density
+
+    def integrate(e):  # define a helper function
+        return trapezoid(e, x=e.T.index, axis=-1)
+
+    int_si = integrate(si)  # integrate spectral irradiance wrt wavelength
+    int_pfd = integrate(pfd)  # integrate photon flux density wrt wavelength
+
+    ape = (1/constants.elementary_charge)*int_si/int_pfd
```

Comment:
Good point.
`ape = (1/constants.elementary_charge)*int_si/int_pfd` returns `np.nan` along with a `RuntimeWarning` because you would are trying to divide something (a zero) by zero.

If the spectral irradiance is zero, I am in favour of returning `np.nan` as the value of APE rather than zero.
Maths: division by zero => output NaN and run away
Physics: While 0 spectral irradiance is a defined measurable condition, it means there are no photons at any wavelength hence there is no average energy of photons to be calculated (not that there are photons with zero energy...)

What do you think?

```
from pvlib import spectrum
import pandas as pd

series = pd.Series(0, range(300, 1201, 5))
spectrum.average_photon_energy(pd.Series(0))
C:\Users\ezxrd3\Documents\GitHub\pvlib-python\pvlib\spectrum\mismatch.py:1190: RuntimeWarning: invalid value encountered in scalar divide
  ape = (1/constants.elementary_charge)*int_si/int_pfd
Out[1]: nan
```

Reply:
Sticking with NaN sounds good to me.  I suggest we use [with np.errstate(...):](https://numpy.org/doc/stable/reference/generated/numpy.errstate.html) to suppress the RuntimeWarning.

Reply:
Do the docs need to be updated to explain this behaviour?

Reply:
A brief sentence in the description in the `Returns` section would be appropriate, I think :)